src/Diagnostics.DataProviders/DataProviders/KustoDataProvider.cs

```diff
@@ -4,26 +4,120 @@
         }
+
+        private void AddQueryInformationToMetadata(string query, string stampName)
+        {
+            KustoQuery q = new KustoQuery
+            {
+                Text = query,               
+                Url = CreateKustoQueryUri(stampName, query)
+            };
+
+            Metadata.PropertyBag.Add(new KeyValuePair<string, object>("Query", q));
+        }
+
+        private string ParseRegionFromStamp(string stampName)
+        {
+            if (string.IsNullOrWhiteSpace(stampName))
+            {
+                throw new ArgumentNullException("stampName");
+            }
+
+            var stampParts = stampName.Split(new char[] { '-' });
+            if (stampParts.Any() && stampParts.Length >= 3)
+            {
+                return stampParts[2];
+            }
+
+            //return * for private stamps if no prod stamps are found
+            return "*";
+        }
+
+        private string CreateKustoQueryUri(string stampName, string query)
+        {
+            string kustoClusterName = null;
+            try
+            {
+                string appserviceRegion = ParseRegionFromStamp(stampName);
+
+                if (!_configuration.RegionSpecificClusterNameCollection.TryGetValue(appserviceRegion.ToLower(), out kustoClusterName))
+                {
+                    if (!_configuration.RegionSpecificClusterNameCollection.TryGetValue("*", out kustoClusterName))
+                    {
+                        throw new KeyNotFoundException(String.Format("Kusto Cluster Name not found for Region : {0}", appserviceRegion.ToLower()));
+                    }
+                }
+
+                string encodedQuery = EncodeQueryAsBase64Url(query);
+
+                var url = string.Format("https://web.kusto.windows.net/clusters/{0}.kusto.windows.net/databases/{1}?q={2}", kustoClusterName, _configuration.DBName, encodedQuery);
+
+                return url;
+            }
+            catch (Exception ex)
+            {
+                string message = string.Format("stamp : {0}, kustoClusterName : {1}, Exception : {2}",
+                    stampName ?? "null",
+                    kustoClusterName ?? "null",
+                    ex.ToString());
+                throw;
+            }
+        }
+
+        // From Kusto.Data.Common.CslCommandGenerator.EncodeQueryAsBase64Url
+        private string EncodeQueryAsBase64Url(string query)
+        {
+            if (string.IsNullOrWhiteSpace(query))
+            {
+                return string.Empty;
+            }
+            byte[] bytes = Encoding.UTF8.GetBytes(query);
+            string result;
+            using (MemoryStream memoryStream = new MemoryStream())
+            {
+                using (GZipStream gZipStream = new GZipStream(memoryStream, CompressionMode.Compress, true))
+                {
+                    gZipStream.Write(bytes, 0, bytes.Length);
```

Comment:
Check out GzipStream.WriteAsync(..)

https://msdn.microsoft.com/en-us/library/system.io.compression.gzipstream.writeasync(v=vs.110).aspx

Reply:
So this method is taken *as is* from KUSTO library. I also think that if this method changes in KUSTO lib, we should just be able to copy it and paste it here in future. Though I understand that using Async would be better but given the fact that it is just the Query Text we are converting to the Gzip, I don't think it is going to be to big so I am ok to leave it as is. Let me know if you think otherwise and then I can change the method.

Reply:
ok.. np.